src/freertos_drivers/esp32/Esp32WS2812.cxx

```diff
@@ -0,0 +1,311 @@
+        ledBufferSize_, ledCount_);
+    // Allocate the LED color data buffer
+    ledDataBuffer_ = (uint8_t *)calloc(1, ledBufferSize_);
+    HASSERT(ledDataBuffer_ != NULL);
+
+    rmt_config_t config = RMT_DEFAULT_CONFIG_TX(pin_, channel_);
+    // reconfigure the clock to 40Mhz
+    config.clk_div = 2;
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Configuring RMT using GPIO %d", channel_,
+        pin_);
+    ESP_ERROR_CHECK(rmt_config(&config));
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Installing RMT driver", channel_);
+    ESP_ERROR_CHECK(rmt_driver_install(channel_, 0, 0));
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Installing RMT translator", channel_);
+    ESP_ERROR_CHECK(rmt_translator_init(channel_, ws2812_rmt_adapter));
+
+    if (timeZeroHighTicks == 0 || timeZeroLowTicks == 0 ||
+        timeOneHighTicks == 0 || timeOneLowTicks == 0)
+    {
+        // convert the number of microseconds for high/low signals to the
+        // number of RMT ticks required.
+        uint32_t counter_clk_hz = 0;
+        ESP_ERROR_CHECK(rmt_get_counter_clock(channel_, &counter_clk_hz));
+        float ratio = (float)counter_clk_hz / 1e9;
+        timeZeroHighTicks = (uint32_t)(ratio * TIME_ZERO_HIGH_NSEC);
+        timeZeroLowTicks = (uint32_t)(ratio * TIME_ZERO_LOW_NSEC);
+        timeOneHighTicks = (uint32_t)(ratio * TIME_ONE_HIGH_NSEC);
+        timeOneLowTicks = (uint32_t)(ratio * TIME_ONE_LOW_NSEC);
+        LOG(VERBOSE,
+            "ESP-WS2812: ratio: %6.4f, T0H:%d, T0L:%d, T1H:%d, T1L:%d",
+            ratio, timeZeroHighTicks, timeZeroLowTicks, timeOneHighTicks,
+            timeOneLowTicks);
+    }
+
+    if (ledCount_ > 1)
+    {
+        LOG(VERBOSE, "ESP-WS2812(%d): Starting background update thread",
+            channel_);
+        // start the background update thread with the name as: rmt.{channel}
+        char name[10];
+        strcpy(name, "rmt.");
+        name[4] = '0' + channel_;
+        name[5] = '\0';
+        start(name, TASK_PRIORITY, 2048);
+    }
+    clear();
+}
+
+void Esp32WS2812::set_led_color(const size_t index, const uint8_t red,
+                                const uint8_t green, const uint8_t blue,
+                                bool update, const uint32_t timeout_ms)
+{
+    HASSERT(index < ledCount_);
+
+    // Check if we are updating the LED color, if not then we can skip the
+    // update.
+    if (ledDataBuffer_[(index * 3) + 0] != green ||
+        ledDataBuffer_[(index * 3) + 1] != red ||
+        ledDataBuffer_[(index * 3) + 2] != blue)
+    {
+        LOG(VERBOSE,
+            "ESP-WS2812(%d): Setting LED(%zu) to (r:%d,g:%d,b:%d)", channel_,
+            index, red, green, blue);
+        // TODO: Add conversion for color orders other than GRB
+        ledDataBuffer_[(index * 3) + 0] = green & 0xFF;
+        ledDataBuffer_[(index * 3) + 1] = red & 0xFF;
+        ledDataBuffer_[(index * 3) + 2] = blue & 0xFF;
+        if (ledCount_ > 1)
+        {
+            sem_.post();
+        }
+        else if (update)
```

Comment:
what's the usecase for update == false? It looks like that will never ever go out as there is no background thread and this call will not send it out.

Reply:
There probably isn't much of a use case for that actually. When there is more than one LED a background thread will be created automatically, when there is only one LED you likely want it to update immediately.